File: testcase2.py
import json

# Corrected sample JSON input without extra braces
json_data = {
    "keys": {
        "n": 10,
        "k": 7
    },
    "1": {
        "base": "6",
        "value": "13444211440455345511"
    },
    "2": {
        "base": "15",
        "value": "aed7015a346d635"
    },
    "3": {
        "base": "15",
        "value": "6aeeb69631c227c"
        
    },
    "4": {
        "base": "16",
        "value": "e1b5e05623d881f"
    },
    "5": {
        "base": "8",
        "value": "316034514573652620673"
    },
    "6": {
        "base": "3",
        "value": "2122212201122002221120200210011020220200"
    },
    "7": {
        "base": "3",
        "value": "20120221122211000100210021102001201112121"
    },
    "8": {
        "base": "6",
        "value": "20220554335330240002224253"
    },
    "9": {
        "base": "12",
        "value": "45153788322a1255483"
    },
    "10": {
        "base": "7",
        "value": "1101613130313526312514143"
    }
}

# Step 1: Extract roots x_i from JSON and convert to decimal integers
roots = []
for key in json_data:
    if key == "keys":
        continue
    base = int(json_data[key]["base"])
    val_str = json_data[key]["value"]
    x_val = int(val_str, base)
    roots.append(x_val)

# Here for example, we pick first k roots for polynomial fit
k = json_data["keys"]["k"]
x_points = roots[:k]

# Step 2: Define corresponding y_i values (replace with your actual y values)
# For example purposes only, defining arbitrary y points:
y_points = [10 * (i + 1) for i in range(k)]  # [10, 20, 30, 40, 50, 60, 70]

print("Using points:")
for i in range(k):
    print(f"(x{i}, y{i}) = ({x_points[i]}, {y_points[i]})")

# Step 3: Calculate denominator for Lagrange basis polynomials
def denominator(xi, xj, xk):
    return (xi - xj) * (xi - xk)

d0 = denominator(x_points[0], x_points[1], x_points[2])
d1 = denominator(x_points[1], x_points[0], x_points[2])
d2 = denominator(x_points[2], x_points[0], x_points[1])

# Step 4: Calculate Lagrange basis polynomial coefficients
def lagrange_basis_coefficients(xi, xj, xk):
    denom = (xi - xj) * (xi - xk)
    a = 1 / denom
    b = - (xj + xk) / denom
    c = (xj * xk) / denom
    return a, b, c

a0, b0, c0 = lagrange_basis_coefficients(x_points[0], x_points[1], x_points[2])
a1, b1, c1 = lagrange_basis_coefficients(x_points[1], x_points[0], x_points[2])
a2, b2, c2 = lagrange_basis_coefficients(x_points[2], x_points[0], x_points[1])

# Step 5: Calculate final polynomial coefficients
a = y_points[0] * a0 + y_points[1] * a1 + y_points[2] * a2
b = y_points[0] * b0 + y_points[1] * b1 + y_points[2] * b2
c = y_points[0] * c0 + y_points[1] * c1 + y_points[2] * c2

print("\nComputed polynomial coefficients:")
print(f"a = {a}")
print(f"b = {b}")
print(f"c = {c}")
print(f"\nPolynomial: P(x) = {a} x^2 + {b} x + {c}")

File: testcase2.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <cmath>
#include <cctype>

// Convert single digit (char) to int value
int charToInt(char c) {
    if (c >= '0' && c <= '9') return c - '0';
    else if (c >= 'a' && c <= 'f') return 10 + (c - 'a');
    else if (c >= 'A' && c <= 'F') return 10 + (c - 'A');
    return -1;
    
}

// Convert string number in given base to decimal (unsigned long long)
unsigned long long convertToDecimal(const std::string& value, int base) {
    unsigned long long result = 0;
    for (char c : value) {
        int val = charToInt(c);
        if (val < 0 || val >= base) {
            std::cerr << "Invalid digit '" << c << "' for base " << base << std::endl;
            exit(1);
        }
        result = result * base + val;
    }
    return result;
}

// Calculate denominator (xi - xj)*(xi - xk)
double denominator(double xi, double xj, double xk) {
    return (xi - xj) * (xi - xk);
}

// Calculate Lagrange basis polynomial coefficients a,b,c for L_i(x)
void lagrange_basis_coefficients(double xi, double xj, double xk, double &a, double &b, double &c) {
    double denom = denominator(xi, xj, xk);
    a = 1.0 / denom;
    b = - (xj + xk) / denom;
    c = (xj * xk) / denom;
}

int main() {
    // Mimic JSON input data in map: key -> pair(base, number string)
    std::map<std::string, std::pair<int, std::string>> json_data = {
        {"1", {6, "13444211440455345511"}},
        {"2", {15, "aed7015a346d635"}},
        {"3", {15, "6aeeb69631c227c"}},
        {"4", {16, "e1b5e05623d881f"}},
        {"5", {8, "316034514573652620673"}},
        {"6", {3, "2122212201122002221120200210011020220200"}},
        {"7", {3, "20120221122211000100210021102001201112121"}},
        {"8", {6, "20220554335330240002224253"}},
        {"9", {12, "45153788322a1255483"}},
        {"10", {7, "1101613130313526312514143"}}
    };

    int k = 7; // Number of points to use for fitting

    // Extract first k roots converted to decimal
    std::vector<double> x_points;
    for (int i = 1; i <= 10 && (int)x_points.size() < k; ++i) {
        std::string key = std::to_string(i);
        if (json_data.find(key) == json_data.end()) continue;
        int base = json_data[key].first;
        std::string val_str = json_data[key].second;
        unsigned long long x_val = convertToDecimal(val_str, base);
        x_points.push_back(static_cast<double>(x_val));
    }

    // Example y values corresponding to x_points (replace with your real values)
    std::vector<double> y_points(k);
    for (int i = 0; i < k; ++i)
        y_points[i] = 10.0 * (i + 1);

    std::cout << "Using points (x,y):\n";
    for (int i = 0; i < k; ++i)
        std::cout << "(x" << i << ", y" << i << ") = (" << x_points[i] << ", " << y_points[i] << ")\n";

    // Compute denominator terms for first three points (quadratic)
    double d0 = denominator(x_points[0], x_points[1], x_points[2]);
    double d1 = denominator(x_points[1], x_points[0], x_points[2]);
    double d2 = denominator(x_points[2], x_points[0], x_points[1]);

    // Compute Lagrange basis coefficients a,b,c for first three points
    double a0, b0, c0, a1, b1, c1, a2, b2, c2;
    lagrange_basis_coefficients(x_points[0], x_points[1], x_points[2], a0, b0, c0);
    lagrange_basis_coefficients(x_points[1], x_points[0], x_points[2], a1, b1, c1);
    lagrange_basis_coefficients(x_points[2], x_points[0], x_points[1], a2, b2, c2);

    // Compute final polynomial coefficients
    double a = y_points[0]*a0 + y_points[1]*a1 + y_points[2]*a2;
    double b = y_points[0]*b0 + y_points[1]*b1 + y_points[2]*b2;
    double c = y_points[0]*c0 + y_points[1]*c1 + y_points[2]*c2;

    std::cout << "\nComputed polynomial coefficients:\n";
    std::cout << "a = " << a << "\n";
    std::cout << "b = " << b << "\n";
    std::cout << "c = " << c << "\n";

    std::cout << "\nPolynomial: P(x) = " << a << " x^2 + " << b << " x + " << c << "\n";

    return 0;
}
